compiler/haskell-ide-core/src/Development/IDE/State/Service.hs

```diff
@@ -100,14 +101,32 @@ setProfiling opts shakeOpts =
 shutdown :: IdeState -> IO ()
 shutdown = shakeShut
 
--- | Run a single action using the supplied service.
+-- | Run a single action using the supplied service. See `runActions`
+-- for more details.
 runAction :: IdeState -> Action a -> IO a
 runAction service action = head <$> runActions service [action]
 
 -- | Run a list of actions in parallel using the supplied service.
+-- This will return as soon as the results of the actions are
+-- available.  There might still be other rules running at this point,
+-- e.g., the ofInterestRule.
 runActions :: IdeState -> [Action a] -> IO [a]
-runActions x = join . shakeRun x
-
+runActions x acts = do
+    var <- newEmptyMVar
+    _ <- shakeRun x acts (putMVar var)
```

Comment:
I'd recommend using a Barrier from Control.Concurrent.Extra as that catches some ways this can go wrong

Reply:
For the moment you might as well use the return value rather than the callback - but I assume you're going to change that next?

Reply:
Switched to a barrier. Using the return value is not the same. `shakeRun` returns an `IO (IO [a])` and if you try to `join` that you end up waiting for all rules to finish.

I also renamed it to `runActionSync`. Given that I managed to mistype this several times myself while making this PR, that definitely seems like a good idea :)